### PR TITLE
Adds win_pagefile module

### PR DESCRIPTION
**Pull request by LiranNis**
##### SUMMARY

Adds module that allows management of **Windows** pagefiles.

##### ISSUE TYPE

 - New Module Pull Request

##### COMPONENT NAME

module: lib/ansible/modules/windows/win_pagefile.ps1
doc: lib/ansible/modules/windows/win_pagefile.py

##### ANSIBLE VERSION

```
ansible 2.3.1.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/etc/ansible/modules/']
  python version = 2.7.5 (default, Aug  2 2016, 04:20:16) [GCC 4.8.5 20150623 (Red Hat 4.8.5-4)]
```

##### ADDITIONAL INFORMATION

```

```

### CODE REVIEW

File: lib/ansible/modules/windows/win_pagefile.py
Comment (jhawkesworth) [inline lib/ansible/modules/windows/win_pagefile.py]: please can you use key: value rather than key=value for the examples.  Sorry if this sounds fussy but key: value is standard now I believe.

**Comment by jhawkesworth** [comment]
Thanks for this.

I have a couple of suggestions (as well as the minor thing about the examples above).

Please could you implement 'check_mode'.  There are now lots of examples as most of the windows modules now support it?

Also, typically modules don't have a 'query' mode.  Instead the usual thing to do is to gather facts.  I don't think there are currently any pagefile facts (I haven't looked), but it would make sense to add them to me.  Obviously that would be a separate pull request.

We are trying to get integration tests for all the windows modules at the moment.  Since you have got a good collection of examples it might not be much more work to create integration tests (which are just playbooks).  Have a look at the test/integration/targets/win* for some examples.  Might make sense to make that into a separate pull request.

Looks good otherwise though, thank you for taking the time to submit this.

**Comment by dagwieers** [comment]
We would like to have integration tests for all (Windows) modules, but in this case I don't know how this will affect our VMs. If it's possible, please add integration tests !

File: lib/ansible/modules/windows/win_pagefile.ps1
Comment (dagwieers) [inline lib/ansible/modules/windows/win_pagefile.ps1]: Please move this up, to the beginning of the code.
I am also wondering if this is really required now that the powershell.ps1 includes it.
cc: @nitzmahone @jborean @jhawkesworth Anyone ? :-)

File: lib/ansible/modules/windows/win_pagefile.ps1
Comment (dagwieers) [inline lib/ansible/modules/windows/win_pagefile.ps1]: I would move this after the parameter handling section. But that's a personal preference.

File: lib/ansible/modules/windows/win_pagefile.py
Comment (dagwieers) [inline lib/ansible/modules/windows/win_pagefile.py]: So booleans require the following statements instead if the above:
```yaml
  type: bool
  default: 'yes'
```
And this will do the required to ensure the documentation is consistent.
We prefer 'yes' and 'no' for booleans as it is the most human readable format.
(And we have to make it strings, otherwise YAML will make it True/False again)

File: lib/ansible/modules/windows/win_pagefile.py
Comment (dagwieers) [inline lib/ansible/modules/windows/win_pagefile.py]: So booleans require the following statements instead if the above:
```yaml
  type: bool
  default: 'no'
```

File: lib/ansible/modules/windows/win_pagefile.py
Comment (dagwieers) [inline lib/ansible/modules/windows/win_pagefile.py]: So booleans require the following statements instead if the above:
```yaml
  type: bool
```

File: lib/ansible/modules/windows/win_pagefile.py
Comment (dagwieers) [inline lib/ansible/modules/windows/win_pagefile.py]: So booleans require the following statements instead if the above:
```yaml
  type: bool
  default: 'no'
```

File: lib/ansible/modules/windows/win_pagefile.py
Comment (dagwieers) [inline lib/ansible/modules/windows/win_pagefile.py]: Please turn this into a list, without the quotes (as we tend to only add quotes when they are required by YAML).
So it should become:
```yaml
author:
- Liran Nisanov (@LiranNis)
```

File: lib/ansible/modules/windows/win_pagefile.py
Comment (dagwieers) [inline lib/ansible/modules/windows/win_pagefile.py]: If it is not required, you don't have to add `required: false`. This is implicit (and makes it easier to see what is required).

**Review by dagwieers — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/windows/win_pagefile.ps1
Comment (LiranNis) [reply lib/ansible/modules/windows/win_pagefile.ps1]: I started to write this module when it wasn't part of powershell.ps1 :) I will remove that.

**Review by dagwieers — APPROVED** [review]

**Comment by LiranNis** [comment]
@jhawkesworth, I added integration tests and check_mode. 
About the query part, after it will be added to the facts (I have few things to add to there) we can remove the query and return pagefiles for every type of task, as part of the result variable, it will require to change the integration tests as well.

The virtual machines of the integration tests have only C drive? I wrote it to check only one pagefile over C drive.
Another question about the vms, you create a new vm (or revert to snapshot) for every run of the tests? I need to know if I need to return the pagefiles to automatic if you don't do so (someone wrote comment about that and deleted it, so I am not sure).

File: lib/ansible/modules/windows/win_pagefile.ps1
Comment (dagwieers) [inline lib/ansible/modules/windows/win_pagefile.ps1]: Could you add here -WhatIf support ?

```powershell
Function Remove-Pagefile($path, $whatif) {
    Get-WmiObject Win32_PageFileSetting | WHERE { $_.Name -eq $path } | Remove-WmiObject -WhatIf:$whatif
}
```

File: lib/ansible/modules/windows/win_pagefile.ps1
Comment (dagwieers) [inline lib/ansible/modules/windows/win_pagefile.ps1]: I think you could do here:

```powershell
$currentPageFiles | Remove-WmiObject -WhatIf:$check_mode | Out-Null
```

And avoid the conditional.

File: lib/ansible/modules/windows/win_pagefile.ps1
Comment (dagwieers) [inline lib/ansible/modules/windows/win_pagefile.ps1]: This would become:
```powershell
Remove-Pagefile $fullPath -WhatIf:$check_mode
```

File: lib/ansible/modules/windows/win_pagefile.ps1
Comment (dagwieers) [inline lib/ansible/modules/windows/win_pagefile.ps1]: This would become:
```powershell
Remove-Pagefile $fullPath -WhatIf:$check_mode
```

**Review by dagwieers — CHANGES_REQUESTED** [review]
Good work !

**Comment by jhawkesworth** [comment]
@LiranNis - right now the integration tests run on amazon - I don't know the details of the AMIs used but I guess it is safest to assume only a C: drive.  

I think safest thing to do is restore the page file state to whatever it was before the integration tests started.  While this might be slower to run, it means the tests can be re-run without having to reset the machine state.

File: lib/ansible/modules/windows/win_pagefile.ps1
Comment (dagwieers) [inline lib/ansible/modules/windows/win_pagefile.ps1]: Forgive me for being pedantic, but here again I would prefer a `-WhatIf:$check_mode` for Set-WmiInstance.
Obviously the rest of the block does still need the `if (-not $check_mode)`.

The reason why we prefer the `-WhatIf` functionality is that it is likely that the function also does more than just skip it based on the boolean. We expect that where possible the function also verifies other conditions within the call.

**Review by dagwieers — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/windows/win_pagefile.py
Comment (dagwieers) [inline lib/ansible/modules/windows/win_pagefile.py]: And just because we show signs of OCD. The items in `description:` should end with a dot.
This is not true for `short_description:`, there we do not allow a dot.

Isn't life beautiful like that ? :-P

**Review by dagwieers — CHANGES_REQUESTED** [review]

File: test/integration/targets/win_pagefile/tasks/main.yml
Comment (dagwieers) [inline test/integration/targets/win_pagefile/tasks/main.yml]: Just a FYI, this can be pure YAML too:
```yaml
- name: Set fact for pagefile expected result
  set_fact:
    expected:
      pagefiles:
      - caption: C:\ 'pagefile.sys'
        description: "'pagefile.sys' @ C:\\"
        initial_size: 1024
        maximum_size: 2048
        name: C:\pagefile.sys
```

Comment (LiranNis) [reply test/integration/targets/win_pagefile/tasks/main.yml]: Thanks, I didn't know that, it's a lot easier :)

**Comment by LiranNis** [comment]
To debug the first error I am downloading and installing Windows Server 2008 SP2 right now.

The second error caused by a weird type of drive, not sure what type is Z: drive, but you can't use Test-Path on it, I can remove this condition but it may cause setting up pagefiles on CD drives for example.

**Comment by mattclay** [comment]
CI failure in the `win_pagefile` test on 2008 SP2 for task `Return all previous pagefiles settings`:
```
{
"changed": true, 
"failed": true, 
"msg": "All items completed"
}
```

File: lib/ansible/modules/windows/win_pagefile.ps1
Comment (jborean93) [inline lib/ansible/modules/windows/win_pagefile.ps1]: You need to enclose the exception part in brackets like `$($_.Exception.Message)`

File: lib/ansible/modules/windows/win_pagefile.ps1
Comment (jborean93) [inline lib/ansible/modules/windows/win_pagefile.ps1]: You need to enclose the exception part in brackets like `$($_.Exception.Message)`

File: lib/ansible/modules/windows/win_pagefile.ps1
Comment (jborean93) [inline lib/ansible/modules/windows/win_pagefile.ps1]: You need to enclose the exception part in brackets like `$($_.Exception.Message)`

File: lib/ansible/modules/windows/win_pagefile.ps1
Comment (jborean93) [inline lib/ansible/modules/windows/win_pagefile.ps1]: You need to enclose the exception part in brackets like `$($_.Exception.Message)`

File: lib/ansible/modules/windows/win_pagefile.ps1
Comment (jborean93) [inline lib/ansible/modules/windows/win_pagefile.ps1]: You need to enclose the exception part in brackets like `$($_.Exception.Message)`

File: lib/ansible/modules/windows/win_pagefile.py
Comment (jborean93) [inline lib/ansible/modules/windows/win_pagefile.py]: Are you able to specify whether this is in bytes, kilobytes, megabytes etc?

File: lib/ansible/modules/windows/win_pagefile.py
Comment (jborean93) [inline lib/ansible/modules/windows/win_pagefile.py]: Are you able to specify whether this is in bytes, kilobytes, megabytes etc?

**Review by jborean93 — CHANGES_REQUESTED** [review]
Looks good, thanks for the contribution @LiranNis, I've just added a few comments.

**Comment by LiranNis** [comment]
@jborean93 thanks for the comments, fixed it.
I am still trying to understand why there is an error in the 2008 SP2 server.

**Comment by jhawkesworth** [comment]
That's an odd one.  Looks like others have hit this:

https://www.vistax64.com/powershell/189685-adding-new-pagefile-another-drive-than-system-drive.html

**Comment by LiranNis** [comment]
From what I saw the VMs don't have Z drive, I added a work around to return it to the way it was (test_path variable).
In the case of the 2008 SP2 server there is not enough free space in C drive so the pagefile can't be set as it was before.
Reference: https://social.technet.microsoft.com/Forums/windows/en-US/46acf090-1da0-42b2-8a37-5628d7fc8068/disabled-automatic-pagefile-management-now-have-pagefile-on-every-volume?forum=winservercore

File: lib/ansible/modules/windows/win_pagefile.ps1
Comment (dagwieers) [inline lib/ansible/modules/windows/win_pagefile.ps1]: This try-catch block is rather large.
I would put 2 different try-catch block around `Set-WmiInstance` and around `$pagefile.Put()`.
Everything else shouldn't fail, and we get more granular error-handling.

File: lib/ansible/modules/windows/win_pagefile.ps1
Comment (dagwieers) [inline lib/ansible/modules/windows/win_pagefile.ps1]: Same here, try to use try-catch for as few statements as possible.

**Review by dagwieers — CHANGES_REQUESTED** [review]

**Comment by LiranNis** [comment]
@dagwieers thanks for the advice, I hope it's better now.

**Comment by dagwieers** [comment]
@LiranNis Perfect ! Much better. We don't have the tendency to catch only specific exceptions in the Windows world (i.e. I haven't seen any Windows module do this correctly), so I am not going to request that. (It's also not always obvious in what ways an instruction can fail, so...)

**Review by dagwieers — APPROVED** [review]
LGTM

**Comment by dagwieers** [comment]
+label new_module

**Comment by LiranNis** [comment]
@jhawkesworth thanks! I decided to use the trick in your link as a workaround, because I still prefer to use WMI functions if I can. 
I tried other ways, most of them caused WMI Provider to crash, so that is the only way.
Microsoft doesn't allow to set these pagefile values through WMI (at least in 2008 SP2), but there is no check through the registry.

Do you think this workaround is ok? Maybe there needs to be a flag to allow users that don't want to use this workaround and prefer fail?
BTW, Can someone please check why there is a pagefile on drive Z when there is no drive Z on the server?

**Comment by LiranNis** [comment]
@jborean93 I fixed the changes you asked for, can you review it?

**Comment by dagwieers** [comment]
So the module freeze deadline for v2.4 is getting closer (2017-08-29). Now is the time to finish up, get it reviewed and merged with no delay.

**Comment by LiranNis** [comment]
@dagwieers in my opinion the module is finished, I am waiting for more reviews.

File: lib/ansible/modules/windows/win_pagefile.ps1
Comment (dagwieers) [inline lib/ansible/modules/windows/win_pagefile.ps1]: So, quite recently there was a change to how we add the GPL license to each individual file. It was changed to only add the following one-liner:
```powershell
# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
```

File: lib/ansible/modules/windows/win_pagefile.py
Comment (dagwieers) [inline lib/ansible/modules/windows/win_pagefile.py]: Same here, the only line remaining (after the Copyright statement(s) is:
```yaml
# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
```

**Review by dagwieers — CHANGES_REQUESTED** [review]

**Comment by dagwieers** [comment]
@LiranNis Reading back over the discussions, I think it would be best to remove the query-information from the module. And move those bits in the default Windows facts. It is easier to leave it out, then to remove it later (as we have to deprecate it for 4 versions...).

@nitzmahone @jborean93 @jhawkesworth So this is about to be merged. Any last remarks ?

**Review by jborean93 — APPROVED** [review]
I'm good to merge this and get it in before the 2.4 freeze.

**Comment by LiranNis** [comment]
@dagwieers I changed the license, if I will remove the query part I won't be able to get the current pagefiles and return it to the same values at the end of the integration run,
And I did few more checks in the integration with it, which means the integration will only check changed / ok / failed and not the values itself, when it will be in facts that is probably what I will do.
I still want to add pagefiles and disks to the facts, I will work on it soon.

**Comment by genadipost** [comment]
@dagwieers This module can be very useful for me (and many others). What holds this merge request from getting in before 2.4 freeze? (and how can i help)

**Comment by dagwieers** [comment]
@genadipost I expect that most open Windows module PRs will be merged before 29/8 (the v2.4 freeze). I don't see a reason to not merge this, but leave the final decision and merge to @jhawkesworth @jborean93 @nitzmahone . Who have a better trackrecord with Windows than me ;-)

**Review by dagwieers — APPROVED** [review]
LGTM

**Comment by LiranNis** [comment]
@jhawkesworth @jborean93 @nitzmahone is it going to be merged in v2.4?

**Comment by jborean93** [comment]
Let's get this in before the freeze, we can deal with bugs in that phase if any are encountered.

**Pull request merged**
